src/types.ts

```diff
@@ -3488,6 +3488,139 @@ export class ZodNaN extends ZodType<number, ZodNaNDef> {
+
+type ZodFileCheck =
+  | { kind: "max"; value: number; message?: string }
+  | { kind: "min"; value: number; message?: string }
+  | { kind: "type"; value: string | string[]; message?: string };
+
+export interface ZodFileDef extends ZodTypeDef {
+  typeName: ZodFirstPartyTypeKind.ZodFile;
+  checks: ZodFileCheck[];
+}
+
+export class ZodFile<FileConstructor extends Function> extends ZodType<
+  FileConstructor,
+  ZodFileDef
+> {
+  constructor(def: ZodFileDef, private readonly fileConstructor: Function) {
+    super(def);
+  }
+
+  _parse(input: ParseInput): ParseReturnType<FileConstructor> {
+    const { status, ctx } = this._processInputParams(input);
+
+    if (!(input.data instanceof this.fileConstructor)) {
+      addIssueToContext(ctx, {
+        code: ZodIssueCode.invalid_type,
+        expected: ZodParsedType.file,
+        received: ctx.parsedType,
+      });
+
+      return INVALID;
+    }
+
+    for (const check of this._def.checks) {
+      if (check.kind === "max") {
+        if (ctx.data.size > check.value) {
+          addIssueToContext(ctx, {
+            code: ZodIssueCode.too_big,
+            maximum: check.value,
+            type: "file",
+            inclusive: true,
+            message: check.message,
+          });
+          status.dirty();
+        }
+      } else if (check.kind === "min") {
+        if (ctx.data.size < check.value) {
+          addIssueToContext(ctx, {
+            code: ZodIssueCode.too_small,
+            minimum: check.value,
+            type: "file",
+            inclusive: true,
+            message: check.message,
+          });
+          status.dirty();
+        }
+      } else if (check.kind === "type") {
+        const types = Array.isArray(check.value) ? check.value : [check.value];
+
+        if (!types.includes(ctx.data.type)) {
+          addIssueToContext(ctx, {
+            code: ZodIssueCode.invalid_file_type,
+            expected: types,
+            received: ctx.data.type,
+            message: check.message,
+          });
+          status.dirty();
+        }
+      }
+    }
+
+    return { status: status.value, value: ctx.data };
+  }
+
+  _addCheck(check: ZodFileCheck) {
+    return new ZodFile(
+      {
+        ...this._def,
+        checks: [...this._def.checks, check],
+      },
+      this.fileConstructor
+    );
+  }
+
+  max(maxSize: number, message?: errorUtil.ErrMessage) {
+    return this._addCheck({
+      kind: "max",
+      value: maxSize,
+      ...errorUtil.errToObj(message),
+    });
+  }
+
+  min(minSize: number, message?: errorUtil.ErrMessage) {
+    return this._addCheck({
+      kind: "min",
+      value: minSize,
+      ...errorUtil.errToObj(message),
+    });
+  }
+
+  type(accept: string | string[], message?: errorUtil.ErrMessage) {
+    return this._addCheck({
+      kind: "type",
+      value: accept,
+      ...errorUtil.errToObj(message),
+    });
+  }
+
+  static create = <T extends Function = typeof File>(
+    fileConstructor?: T
```

Comment:
If we want to support custom errors I think you need to pass [RawCreateParams](https://github.com/colinhacks/zod/blob/efc37cb94d8d21161529462e4ad3109078ba119c/src/types.ts#L644) like in `ZodString` right?

Reply:
I'm talking about this:
![image](https://user-images.githubusercontent.com/49499/158187672-584ca54b-8520-4370-9b81-b2d6f6c55cf7.png)

Reply:
I think is the zod way to allow translated errors

Reply:
Ok, so no problem? :D 
I have removed some extra `"` in "File not supported ..." error message, also changed separator in `issue.expected.join` from `,` to ` or `

Reply:
nvm, i see what you talking about, will update

Reply:
i wonder which parameter should be first, for now file constructor is second